library/std/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,39 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer
+/// with a different signature, the two signatures must be *ABI-compatible* or else this call is
+/// Undefined Behavior. ABI compatibility is a lot stricter than merely having the same
+/// representation in memory; for example, even if `i32` and `f32` have the same size and alignment,
+/// they might be passed in different registers and hence not be ABI-compatible.
+///
+/// For two signatures to be considered *ABI-compatible*, they must declare the same `extern` ABI
+/// string, must take the same number of arguments, and the individual argument types and the return
+/// types must be ABI-compatible.
+/// The relation of when two types are ABI-compatible is defined as follows:
+///
+/// - Every type is ABI-compatible with itself.
+/// - If `<T as Pointee>::Metadata == ()`, then `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>`,
+///   `NonNull<T>` are all ABI-compatible with each other.
+/// - Any two `fn()` types with the same `extern` ABI string are ABI-compatible with each other.
+/// - Any two 1-ZST types (types with size 0 and alignment 1) are ABI-compatible.
```

Comment:
This is currently *not* true on [sparc64](https://github.com/rust-lang/rust/issues/115336) and [mips64](https://github.com/rust-lang/rust/issues/115404), but this is also required for `repr(transparent)` to work so those are likely target bugs. Also see the discussion on IRLO [here](https://internals.rust-lang.org/t/creating-1-zsts-guaranteed-to-have-same-extern-c-abi-as/19399) and [here](https://internals.rust-lang.org/t/pre-rfc-define-the-behavior-of-repr-transparent-when-all-fields-are-zero-sized-types/19415).